qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1445 @@
+        self.Q = Q
+        self.ck = ck
+        self.vk = vk
+        self.ck2 = ck2
+        self.sigma_bar_k_offset = sigma_bar_k_offset
+
+
+class Bath:
+    """
+    Represents a list of bath expansion exponents.
+
+    Parameters
+    ----------
+    exponents : list of BathExponent
+        The exponents of the correlation function describing the bath.
+
+    Attributes
+    ----------
+
+    All of the parameters are available as attributes.
+    """
+    def __init__(self, exponents):
+        self.exponents = exponents
+
+
+class BosonicBath(Bath):
+    """
+    A helper class for constructing a bosonic bath from the expansion
+    coefficients and frequencies for the real and imaginary parts of
+    the bath correlation function.
+
+    Parameters
+    ----------
+    Q : Qobj
+        The coupling operator for the bath.
+
+    ck_real : list of complex
+        The coefficients of the expansion terms for the real part of the
+        correlation function. The corresponding frequencies are passed as
+        vk_real.
+
+    vk_real : list of complex
+        The frequencies (exponents) of the expansion terms for the real part of
+        the correlation function. The corresponding ceofficients are passed as
+        ck_real.
+
+    ck_imag : list of complex
+        The coefficients of the expansion terms in the imaginary part of the
+        correlation function. The corresponding frequencies are passed as
+        vk_imag.
+
+    vk_imag : list of complex
+        The frequencies (exponents) of the expansion terms for the imaginary
+        part of the correlation function. The corresponding ceofficients are
+        passed as ck_imag.
+
+    combine : bool, default True
+        Whether to combine exponents with the same frequency (and coupling
+        operator). See :meth:`combine` for details.
+    """
+    def _check_cks_and_vks(self, ck_real, vk_real, ck_imag, vk_imag):
+        if len(ck_real) != len(vk_real) or len(ck_imag) != len(vk_imag):
+            raise ValueError(
+                "The bath exponent lists ck_real and vk_real, and ck_imag and"
+                " vk_imag must be the same length."
+            )
+
+    def _check_coup_op(self, Q):
+        if not isinstance(Q, Qobj):
+            raise ValueError("The coupling operator Q must be a Qobj.")
+
+    def __init__(self, Q, ck_real, vk_real, ck_imag, vk_imag, combine=True):
+        self._check_cks_and_vks(ck_real, vk_real, ck_imag, vk_imag)
+        self._check_coup_op(Q)
+
+        exponents = []
+        exponents.extend(
+            BathExponent("R", None, Q, ck, vk)
+            for ck, vk in zip(ck_real, vk_real)
+        )
+        exponents.extend(
+            BathExponent("I", None, Q, ck, vk)
+            for ck, vk in zip(ck_imag, vk_imag)
+        )
+
+        if combine:
+            exponents = self.combine(exponents)
+
+        super().__init__(exponents)
+
+    @classmethod
+    def combine(cls, exponents, rtol=1e-5, atol=1e-7):
```

Comment:
this is much better than the old approach, nice to get real or imaginary parts grouped up as well (rather than just a warning).

Tor the tolerances, I think I chose fiddled with the original values so that they worked well for some specific examples.  I think allowing them to be passed by the user is a good choice.